// ui/component/theme/badge.ts
import BaseComponent from "ui/component/component";

export default class Badge extends BaseComponent
{
    public static template() {
        return `<span class="badge bg-{{bg}} {{class}}">{{label}}</span>`
    }

    public static properties() {
        let props = super.properties();
        props['label'] = {
            label: "Text",
            type: "mmTextfield"
        };
        props['bg'] = {
            label: "Type",
            type: "mmDropdown",
            props: {
                options: [
                    {label: "Primary", value: "primary"},
                    {label: "Secondary", value: "secondary"},
                    {label: "Info", value: "info"},
                    {label: "Success", value: "success"},
                    {label: "Warning", value: "warning"},
                    {label: "Danger", value: "danger"},
                    {label: "Light", value: "light"},
                    {label: "Dark", value: "dark"}
                ]
            },
            default: "primary"
        };

        return props;
    }
}
