// ui/component/theme/breadcrumb.ts
import BaseComponent from "ui/component/component";

export default class Breadcrumb extends BaseComponent 
{
    public static template() {
        return `
            <nav class="{{class}}">
                {{#if links}}
                <ol class="breadcrumb mb-0">
                    {{#each links}}
                    <li class="breadcrumb-item {{echo @last 'active'}}">
                        {{#if @last}}
                        <span>{{this.label}}</span>
                        {{else}}
                        <a href="{{this.href}}">{{this.label}}</a>
                        {{/if}}
                    </li>
                    {{/each}}
                </ol>
                {{/if}}
            </nav>
        `;
    }

    public static properties() {
        let props = super.properties();

        props["links"] = {
            label: "Links",
            type: "mmRepeater",
            props: {
                columns: [
                    {
                        key: "label",
                        label: "Label",
                        component: {
                            type: "mmTextfield"
                        }
                    },
                    {
                        key: "value",
                        label: "Value",
                        primary: true,
                        component: {
                            type: "mmTextfield"
                        }
                    },
                    {
                        key: "href",
                        label: "Href",
                        component: {
                            type: "mmTextfield"
                        }
                    }
                ]
            }
        };

        return props;
    }
}
