// ui/component/theme/button.ts
import BaseComponent from "ui/component/component";

export default class Button extends BaseComponent
{
    public initScope() {
        super.initScope();
        const comp = this;
        
        this.scope.click = function(e) {
            e.preventDefault();
            e.stopPropagation();
            comp.trigger("click");
        }
    }

    public static template() {
        return `
            <button class="position-relative btn{{echo icon (echo label '' ' btn-icon')}} btn-{{bg}} btn-{{size}} {{class}}" type="button" {{echo disabled 'disabled'}} mm-click="this.click(event)">
                {{#if loading}}
                    <span class="spinner-border spinner-border-sm" role="status" aria-hidden="true"></span>
                    Loading...
                {{else}}
                    {{#if icon}}
                    <i class="fa fa-{{icon}}"></i>
                    {{/if}}
                    {{{echo label label}}}
                {{/if}}
                {{#if badge}}
                <span class="badge {{badge.class}}">{{badge.label}}</span>
                {{/if}}
            </button>
        `;
    }
    
    public static properties() {
        let props = super.properties();
        props['label'] = {
            label: "Text",
            type: "mmTextfield"
        };
        props['bg'] = {
            label: "Background",
            type: "mmDropdown",
            props: {
                options: [
                    {label: "Primary", value: "primary"},
                    {label: "Secondary", value: "secondary"},
                    {label: "Info", value: "info"},
                    {label: "Success", value: "success"},
                    {label: "Warning", value: "warning"},
                    {label: "Danger", value: "danger"},
                    {label: "Light", value: "light"},
                    {label: "Dark", value: "dark"},
                    {label: "Primary Outline", value: "outline-primary"},
                    {label: "Secondary Outline", value: "outline-secondary"},
                    {label: "Info Outline", value: "outline-info"},
                    {label: "Success Outline", value: "outline-success"},
                    {label: "Warning Outline", value: "outline-warning"},
                    {label: "Danger Outline", value: "outline-danger"},
                    {label: "Light Outline", value: "outline-light"},
                    {label: "Dark Outline", value: "outline-dark"},
                    {label: "Link", value: "link"}
                ]
            },
            default: "primary"
        };
        props['icon'] = {
            label: "Icon",
            type: "mmTextfield"
        };
        props['size'] = {
            label: "Size",
            type: "mmDropdown",
            props: {
                options: [
                    {label: "Default", value: ""},
                    {label: "Large", value: "lg"},
                    {label: "Small", value: "sm"}
                ]
            },
            default: ""
        };

        props['disabled'] = {
            label: "Disabled",
            type: "mmCheckbox",
            default: false,
            priority: 0
        };

        props["badge"] = {
            label: "Badge",
            type: "mmForm",
            props: {
                elements: [
                    {
                        comp: {
                            name: "bg",
                            type: "mmDropdown",
                            props: {
                                options: [
                                    {label: "Primary", value: "primary"},
                                    {label: "Secondary", value: "secondary"},
                                    {label: "Info", value: "info"},
                                    {label: "Success", value: "success"},
                                    {label: "Warning", value: "warning"},
                                    {label: "Danger", value: "danger"},
                                    {label: "Light", value: "light"},
                                    {label: "Dark", value: "dark"},
                                    {label: "Primary Outline", value: "outline-primary"},
                                    {label: "Secondary Outline", value: "outline-secondary"},
                                    {label: "Info Outline", value: "outline-info"},
                                    {label: "Success Outline", value: "outline-success"},
                                    {label: "Warning Outline", value: "outline-warning"},
                                    {label: "Danger Outline", value: "outline-danger"},
                                    {label: "Light Outline", value: "outline-light"},
                                    {label: "Dark Outline", value: "outline-dark"},
                                    {label: "Link", value: "link"}
                                ]
                            }
                        }
                    }
                ]
            },
            default: ""
        };

        props['loading'] = {
            label: "Loading",
            type: "mmCheckbox",
            default: false
        };

        return props;
    }

    public static events() {
        let events = super.events();
        events['click'] = {
            label: "Button click event"
        };

        return events;
    }

    public setLoading(bool) {
        this.setScopeVal("loading", bool);
        this.setScopeVal("disabled", bool);
    }
}
