// ui/component/theme/heading.ts
import BaseComponent from "ui/component/component";

export default class Heading extends BaseComponent 
{
    public static template() {
        return `
            <div class="{{class}}">
                <h{{size}}>
                    {{#if icon}}
                    <i class="fa fa-{{icon}} mm-heading-icon"></i>
                    {{/if}}
                    {{label}}
                    {{#if badge}}
                    <span class="badge bg-{{badge.type}}">{{badge.label}}</span>
                    {{/if}}
                    <small class="mm-heading-description {{echo badge 'display-block' ''}}">{{{description}}}</small>
                </h{{size}}>
            </div>    
        `;
    }

    public static properties() {
        let props = super.properties();

        props['label'] = {
            label: "Text",
            type: "mmTextfield"
        };
        props['size'] = {
            label: "Title size",
            type: "mmNumber",
            props: {
                min: 1,
                max: 6
            },
            default: 3
        };
        props['badge'] = {
            label: "Badge",
            type: "mmForm",
            props: {
                elements: [
                    {
                        comp: {
                            name: "label",
                            type: "mmTextfield"
                        },
                        option: {
                            label: "Label"
                        }
                    },
                    {
                        comp: {
                            name: "type",
                            type: "mmDropdown",
                            props: {
                                options: [
                                    {label: "Primary", value: "primary"},
                                    {label: "Default", value: "default"},
                                    {label: "Warning", value: "warning"},
                                    {label: "Danger", value: "danger"}
                                ]
                            }
                        },
                        option: {
                            label: "Type"
                        }
                    }
                ]
            },
            schema: {
                type: "object",
                properties: {
                    label: {
                        type: "string"
                    },
                    type: {
                        enum: ["primary", "default", "warning", "danger"]
                    }
                }
            }
        };
        props['description'] = {
            label: "Description",
            type: "mmTextarea"
        };
        props['icon'] = {
            label: "Icon",
            type: "mmTextfield"
        };

        return props;
    }
}
