// ui/component/theme/overlay.ts
import BaseComponent from "ui/component/component";

export default class Overlay extends BaseComponent
{
    public initScope() {
        super.initScope();
        const comp = this;
    }

    public static template() {
        return `
            <div class="{{class}} p-3 bg-light {{echo visible 'd-flex' 'd-none'}} justify-content-center align-items-center" style="--bs-bg-opacity: .5;">
                {{#if loading}}
                <div class="spinner-border" role="status">
                </div>
                {{/if}}
            </div>
        `;
    }
    
    public static properties() {
        let props = super.properties();
        
        props['loading'] = {
            label: "Loading",
            type: "mmCheckbox",
            default: true
        };

        props['visible'] = {
            label: "Visbile",
            type: "mmCheckbox",
            default: false
        };

        props['class']['default'] = "position-fixed"

        return props;
    }

    public display(comp: BaseComponent) {
        const w = comp.element.width();
        const h = comp.element.height();
        const offset = comp.element.offset();

        comp.element.append(this.element);

        this.prop('visible', true);

        this.element.width(w);
        this.element.height(h);
        this.element.offset(offset);
    }
}
