// ui/component/theme/pager.ts
import BaseComponent from "ui/component/component";

export default class Pager extends BaseComponent
{
    public initScope(data = null) {
        super.initScope(data);
        const comp = this;
        this.scope.next = function(e) {
            e.stopPropagation();
            e.preventDefault();
            comp.prop("page", this.page + 1);
            comp.trigger("page");
        }
        this.scope.last = function(e) {
            e.stopPropagation();
            e.preventDefault();
            comp.prop("page", this.total);
            comp.trigger("page");
        }
        this.scope.prev = function(e) {
            e.stopPropagation();
            e.preventDefault();
            comp.prop("page", this.page - 1);
            comp.trigger("page");
        }
        this.scope.first = function(e) {
            e.stopPropagation();
            e.preventDefault();
            comp.prop("page", 1);
            comp.trigger("page");
        }
    }
    
    public static template() {
        return `
            <div class="{{class}} {{echo (lte total 1) 'd-none'}}">
                <ul class="pagination float-{{position}}">
                    <li class="page-item {{echo (equal page 1) 'disabled'}}">
                        <a class="page-link" mm-click="this.first(event)"><i class="fa fa-angle-double-left"></i></a>
                    </li>
                    <li class="page-item {{echo (equal page 1) 'disabled'}}">
                        <a class="page-link" mm-click="this.prev(event)"><i class="fa fa-angle-left"></i></a>
                    </li>
                    <li class="page-item d-flex p-2 align-items-center justify-content-center">
                        <span>{{page}} of {{total}}</span>
                    </li>
                    <li class="page-item {{echo (equal total page) 'disabled'}}">
                        <a class="page-link" mm-click="this.next(event)"><i class="fa fa-angle-right"></i></a>
                    </li>
                    <li class="page-item {{echo (equal total page) 'disabled'}}">
                        <a class="page-link" mm-click="this.last(event)"><i class="fa fa-angle-double-right"></i></a>
                    </li>
                </ul>
            </div>   
        `;
    }

    public static properties() {
        let props = super.properties();
        props['numFound'] = {
            label: "Total",
            type: "mmNumber",
            default: 0
        };
        props['page'] = {
            label: "Page",
            type: "mmNumber",
            default: 1
        };
        props['limit'] = {
            label: "Limit",
            type: "mmNumber",
            default: 20
        };
        props['position'] = {
            label: "Position",
            type:"mmDropdown",
            default: "start",
            props: {
                options: [
                    {label: "Left", value: "start"},
                    {label: "Right", value: "end"}
                ]
            }
        };
        
        return props;
    }

    public static events() {
        let events = super.events();
        events['page'] = {
            label: "Page change event"
        };

        return events;
    }

    public setLimit(limit) {
        if (limit === undefined || ! limit) {
            let props = this.getProperties();
            limit = props["limit"]["default"];
        } else {
            limit = parseInt(limit);
        }
        let total =  Math.ceil(this.prop("numFound") / limit);
        
        this.setScopeVal("limit", limit);
        this.setScopeVal("total", total);
    }

    public setNumFound(numFound) {
        numFound = parseInt(numFound);
        let total =  Math.ceil(numFound / this.prop("limit"));
        this.setScopeVal("numFound", numFound);
        this.setScopeVal("total", total);
    }

    public getTotal() {
        return this.getScopeVal("total");
    }
}
