// ui/component/theme/progress.ts
import BaseComponent from "ui/component/component";

export default class Progress extends BaseComponent
{
    public static template() {
        return `
            <div class="d-flex align-items-center {{class}}">
                <div class="progress flex-grow-1 {{echo (not visible) 'd-none'}}" style="{{echo isLine 'height: 1px'}}">
                    <div class="progress-bar progress-bar-striped progress-bar-animated" style="width:{{perct}}%"></div>
                </div>
                {{#if showLabel}}
                <small class="text-dark ms-1">{{perct}}%</small>
                {{/if}}
            </div>
        `;
    }

    public static properties() {
        let props = super.properties();
        
        props['perct'] = {
            label: "Percentage",
            type: "mmTextfield",
            default: "0"
        };

        props['isLine'] = {
            label: "Line progress bar",
            type: "mmCheckbox",
            default: false
        };

        props['visible'] = {
            label: "Visible",
            type: "mmCheckbox",
            default: false
        };

        props['showLabel'] = {
            label: "Display label",
            type: "mmCheckbox",
            default: true
        };

        return props;
    }
}
