// ui/component/theme/titlebar.ts
import BaseContainer from "ui/component/container";

export default class Titlebar extends BaseContainer
{
    public constructor(options) {
        if (! options.props) {
            options.props = {};    
        }
        options.props.components = {
            left: {
                type: "mmFlexLayout",
                props: {
                    class: "align-items-center justify-content-start"
                }
            },
            right: {
                type: "mmFlexLayout",
                props: {
                    class: "align-items-center justify-content-end"
                }
            }
        };
        super(options);
    }

    public static template() {
        return `
            <div class="p-2 rounded d-flex {{class}}">
                <h5 class="pb-0 mb-0 me-2 d-flex align-items-center">
                    {{#if icon}}
                    <span class="text-dark me-1"><i class="fa fa-{{icon}}"></i></span>
                    {{/if}}
                    {{#if label}}
                    <span>{{label}}</span>
                    {{/if}}
                    {{#if description}}
                    <small class="ms-1 text-black-50">{{description}}</small>
                    {{/if}}
                </h5>
                <div class="flex-grow-1 d-flex justify-content-between">
                    {{#if components.left}}
                    <mm-component name="{{components.left.name}}" type="{{components.left.type}}"></mm-component>
                    {{/if}}
                    {{#if components.right}}
                    <mm-component name="{{components.right.name}}" type="{{components.right.type}}"></mm-component>
                    {{/if}}
                </div>
            </div>
        `;
    }

    public static properties() {
        let props = super.properties();
        props['class']['default'] = 'bg-light';
        props['icon'] = {
            label: "Icon",
            type: "mmTextfield"
        };
        props['label'] = {
            label: "Title",
            type: "mmTextfield",
            default: ""
        };
        props['description'] = {
            label: "Description",
            type: "mmTextfield"
        };
        props['right'] = {
            label: "Right components",
            type: "mmRepeater",
            props: {
                isSingleColumn: true,
                columns: [
                    {
                        key: "comp",
                        primary: true,
                        label: "Component",
                        component: {
                            type: "mmComponentForm"
                        }
                    }
                ]
            },
            priority: 0
        };
        props['left'] = {
            label: "Left components",
            type: "mmRepeater",
            props: {
                isSingleColumn: true,
                columns: [
                    {
                        key: "comp",
                        primary: true,
                        label: "Component",
                        component: {
                            type: "mmComponentForm"
                        }
                    }
                ]
            },
            priority: 0
        };
        
        return props;
    }

    public setRight(components) {
        if (this.isArray(components)) {
            this.get("right").empty();
            this.get("right").prop("components", components);
        }
    }

    public setLeft(components) {
        if (this.isArray(components)) {
            this.get("left").empty();
            this.get("left").prop("components", components);
        }
    }
}
