// ui/component/theme/form/checkbox.ts
import BaseFormElement from "ui/component/form.element";
import * as $ from "jquery/dist/jquery.min.js";

export default class Checkbox extends BaseFormElement
{
    public initScope() {
        super.initScope();
        const comp = this;
        this.scope.click = function(e) {
            e.stopPropagation();
            const value = comp.element.find("input").is(":checked");
            comp.prop("value", value);
            comp.trigger("click");
        }
    }
    public static template() {
        return `
            <div class="form-check{{echo switch ' form-switch'}} {{class}}">
                <input class="form-check-input" type="checkbox" id="{{name}}" {{checked value}} {{disabled disabled}} mm-click="this.click(event)">
                {{#if label}}
                <label class="form-check-label" for="{{name}}">{{{label}}}</label>
                {{/if}}
            </div>
        `;
    }

    public static properties() {
        let props = super.properties();
        
        props["value"] = {
            label: "Value",
            type: "mmCheckbox",
            default: false
        };
        props["switch"] = {
            label: "Switch",
            type: "mmCheckbox",
            default: false
        };
        props["label"] = {
            label: "Label",
            type: "mmTextfield",
            default: ""
        };
        props["disabled"] = {
            label: "Disabled",
            type: "mmCheckbox",
            default: false
        };
        props["trueValue"] = {
            label: "True Value",
            type: "mmTextfield",
            default: true
        };
        props["falseValue"] = {
            label: "False Value",
            type: "mmTextfield",
            default: false
        };

        return props;
    }

    public static events() {
        let events = super.events();

        events["click"] = {
            label: "Click"
        };

        return events;
    }

    public getValue() {
        if (this.getScopeVal("value")) {
            return this.getScopeVal("trueValue");
        }
        return this.getScopeVal("falseValue");
    }
}
