// ui/component/theme/form/function.ts
import BaseFormElement from "ui/component/form.element";

export default class BaseFunction extends BaseFormElement {
    public constructor(options) {
        if (! options.props) {
            options.props = {};
        }
        options.props.components = {
            main: {
                type: "mmGrid",
                props: {
                    components: [
                        {
                            comp: {
                                type: "mmTags"
                            },
                            option: {
                                label: "Parameters"
                            }
                        },
                        {
                            comp: {
                                type: "mmCodeEditor",
                                options: {
                                    language: "javascript"
                                },
                                props: {
                                    height: 300
                                }
                            },
                            option: {
                                label: "Function Body"
                            }
                        }
                    ]
                }
            }
        };

        super(options);
    }

    public setValue(value) {
        const tagInput = this.get("main").get(0);
        const codeInput = this.get("main").get(1);
        if (! value) {
            tagInput.prop("value", "");
            codeInput.prop("value", "");
        } else {
            tagInput.prop("value", value.params);
            codeInput.prop("value", value.body);
        }
    }

    public getValue() {
        const tagInput = this.get("main").get(0);
        const codeInput = this.get("main").get(1);

        return {
            params: tagInput.prop("value"),
            body: codeInput.prop("value")
        };
    }
}
