// ui/component/theme/form/select.ts
import BaseFormElement from "ui/component/form.element";

export default class Select extends BaseFormElement {
    public static template() {
        return `
            <select class="form-select {{size}} {{class}}" {{disabled disabled}}>
                {{#if options}}
                {{#each options}}
                <option value="{{this.value}}" {{echo (equal this.value @root/value) 'selected'}}>{{this.label}}</option>
                {{/each}}
                {{/if}}
            </select>
        `;
    }

    public static properties() {
        let props = super.properties();

        props["size"] = {
            label: "Size",
            type: "mmDropdown",
            props: {
                options: [
                    {label: "Default", value: ""},
                    {label: "Large", value: "form-select-lg"},
                    {label: "Small", value: "form-select-sm"}
                ]
            },
            default: ""
        };

        props["disabled"] = {
            label: "Disabled",
            type: "mmCheckbox",
            default: false
        };

        props["options"] = {
            label: "Options",
            type: "mmRepeater",
            props: {
                columns: [
                    {
                        key: "label",
                        label: "Label",
                        component: {
                            type: "mmTextfield"
                        }
                    },
                    {
                        key: "value",
                        label: "Value",
                        primary: true,
                        component: {
                            type: "mmTextfield"
                        }
                    }
                ]
            }
        };

        return props;
    }
}
