// ui/component/theme/form/textarea.ts
import BaseFormElement from "ui/component/form.element";
import * as $ from "jquery/dist/jquery.min.js";

export default class Textarea extends BaseFormElement
{
    public initScope() {
        super.initScope();
        let comp = this;
        this.scope.typing = function(e) {
            e.stopPropagation();
            this.value = $(e.target).val();
            if (e.keyCode == 13 || e.which == 13) {
                comp.trigger("enter");
            }
            if (comp.changed()) {
                comp.trigger("change", comp.initialValue, comp.prop("value"));
            }
        }
    }
    public static template() {
        return `
            <textarea class="form-control {{class}}" rows="{{rows}}" mm-keyup="this.typing(event)" {{disabled disabled}} {{readonly readonly}}>{{{value}}}</textarea>
        `;
    }

    public static properties() {
        let props = super.properties();

        props["rows"] = {
            label: "Rows",
            type: "mmNumber",
            default: "5"
        };

        props["disabled"] = {
            label: "Disabled",
            type: "mmCheckbox",
            default: false
        };

        props["readonly"] = {
            label: "Readonly",
            type: "mmCheckbox",
            default: false
        };

        return props;
    }

    public static events() {
        let events = super.events();
        events['enter'] = {
            label: "Enter"
        };

        return events;
    }
}
