// ui/component/theme/container/content.ts
import app from "ui/module/index";
import BaseContainer from "ui/component/container";
import BaseComponent from "ui/component/component";

export default class Content extends BaseContainer
{
    public static template() {
        return `<div class="{{class}}" {{echo style (style style)}}>
            {{#if (and isComponent components.content)}}
            <mm-component name="{{components.content.name}}" type="{{components.content.type}}"></mm-component>
            {{/if}}
            {{#if (and (not isComponent) template)}}
            {{{compile template data}}}
            {{/if}}
        </div>`;
    }

    public static properties() {
        let props = super.properties();
        props['style'] = {
            label: "Style",
            type: "mmCodeEditor"
        };
        props['content'] = {
            label: "Content",
            type: "mmTabs",
            props: {
                active: 'html',
                tabs: [
                    {
                        label: "HTML",
                        value: "html",
                        content: {
                            type: "mmTextarea"
                        }
                    },
                    {
                        label: "Component",
                        value: "component",
                        content: {
                            type: "mmComponentForm"
                        }
                    }
                ]
            }
        };
        props['data'] = {
            label: "Data",
            type: "mmCodeEditor"
        };

        return props;
    }
    
    public setContent(content) {
        if (content instanceof BaseComponent) {
            this.setScopeVal("content", content);
            this.setScopeVal("isComponent", true);
            this.add(content, "content");
            return;
        }
        this.setScopeVal("content", this.copy(content));
        if (this.isObject(content)) {
            this.add(app.create(content), "content");
            this.setScopeVal("isComponent", true);
        } else if (this.isString(content)) {
            this.setScopeVal("template", content);
            this.setScopeVal("isComponent", false);
        }
    }

    public getComponentContent() {
        return this.get('content');
    }
}
