// ui/component/theme/layout/flex.ts
import  BaseLayout from "ui/component/layout";

export default class FlexLayout extends BaseLayout
{
    public static template() {
        return `
            <div class="d-flex {{class}}">
                {{#each components}}
                <div class="{{../itemClass}} {{echo this.option.class this.option.class}}">
                    <mm-component name="{{this.comp.name}}" type="{{this.comp.type}}"></mm-component>
                </div>
                {{/each}}
            </div>
        `;
    }

    public static properties() {
        let props = super.properties();
        props["itemClass"] = {
            label: "Class for each flex item",
            type: "mmTags",
            default: ""
        };
        

        return props;
    }

    public setItemClass(classes) {
        if (this.isArray(classes)) {
            this.setScopeVal("itemClass", classes.join(" "));
        } else {
            this.setScopeVal("itemClass", classes);
        }
    }

    public getItemClass() {
        const classes = this.getScopeVal("itemClass");
        if (! classes) {
            return "";
        }
        return classes.split(" ");
    }
}
